### PR TITLE
GH646: Add module support.

### LINKED ISSUE #646: Add Module Support

### PR DESCRIPTION
**Pull request by patriksvensson**
This PR provides module support and the ability to swap out or add new parts to Cake.

- [x] Add abstraction around container.
- [x] Move existing registrations to modules in each assembly.
- [x] Add tests for container stuff in the Cake asembly.
- [x] Load modules at start according to conventions.
- [x] Make convention for module directory customizable via configuration.
- [x] Discover assemblies using reflection-only context.

## Registration rules

* When resolving a type, we will always return the last registered type.
* If resolving an enumerable of a registration type, all implementations will be returned.

```csharp
public interface IFoo { }
public class Bar : IFoo { }
public class Qux : IFoo { }
````

And performs registration of them:

```csharp
public class MyModule : ICakeModule
{
  public void Register(ICakeContainerRegistry registry)
  {
    registry.RegisterType<Bar>().As<IFoo>().Singleton();
    registry.RegisterType<Qux>().As<IFoo>().Singleton();
  }
}
```

* Resolving `IFoo` would return an instance of `Qux`.
* Resolving `IEnumerable<IFoo>` would return both `Bar` and `Qux`.

## Example

Let's say that you want to create a log that reverses everything that is written to it.
Create a new project that you name `Cake.MyAssembly.Module.dll`.

```csharp
public sealed class ReverseLog : ICakeLog
{
    private readonly IConsole _console;
    public Verbosity Verbosity { get; set; }

    public ReverseLog(IConsole console)
    {
        _console = console;
    }

    public void Write(Verbosity verbosity, LogLevel level, string format, params object[] args)
    {
        var message = string.Format(format, args);
        _console.WriteLine(new string(message.Reverse().ToArray()));
    }
}
```

Now you have to create a module in the same assembly that registers the new log with Cake. In the code below, we say that all classes that request a `ICakeLog` should receive a `ReverseLog`. We also tell Cake that there should be only one instance of this class by marking it as a `Singleton`.

```csharp
public class MyModule : ICakeModule
{
    public void Register(ICakeContainerRegistry registry)
    {
        registry.RegisterType<ReverseLog>().As<ICakeLog>().Singleton();
    }
}
```

When Cake looks for modules, it will look for a `CakeModuleAttribute` used in the assembly scope. This attribute should tell Cake what class to instanciate when registering a module.

```csharp
[assembly: CakeModule(typeof(Cake.MyAssembly.Module.MyModule))]
```

Now you copy the assembly to the `tools/modules` folder and run the Cake script like normal. You could also change what folder that Cake should look in by passing `--paths_modules=\"C:/MyModules\"` to Cake, setting an environment variable called `CAKE_PATHS_MODULES` or adding the following to your `cake.config` file.

```ini
[Paths]
Modules=C:/MyModules
```

Closes #646

### CODE REVIEW

**Comment by patriksvensson** [comment]
@cake-build/team This is ready for review.

File: src/Cake.NuGet/NuGetModule.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+﻿using Cake.Core.Composition;
+using Cake.Core.Packaging;
+
+namespace Cake.NuGet
+{
+    /// <summary>
+    /// The module responsible for registering
+    /// default types in the Cake.NuGet assembly.
+    /// </summary>
+    public sealed class NuGetModule : ICakeModule
```
Comment (gep13) [inline src/Cake.NuGet/NuGetModule.cs]: @patriksvensson  Would this make more sense as CakeNugetModule?

Comment (patriksvensson) [inline src/Cake.NuGet/NuGetModule.cs]: Not sure to be honest. We haven't prefixed any other modules with `Cake`.

Comment (gep13) [inline src/Cake.NuGet/NuGetModule.cs]: @patriksvensson  Sorry, this was the first/last comment that I made on my first brief pass through the PR.  I would have likely made the same comment on each of the modules.  I don't have any strong arguments either way on this particular naming, but it just sort of stood out when I was going through it.

File: src/Cake/Composition/ModuleLoader.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                {
+                    // Register modules in the temporary container.
+                    RegisterExternalModules(moduleTypes, temporaryContainer);
+
+                    // Now let the modules register their types.
+                    var builder = new CakeContainerBuilder();
+                    foreach (var module in temporaryContainer.Resolve<IEnumerable<ICakeModule>>())
+                    {
+                        builder.Registry.RegisterModule(module);
+                    }
```
Comment (gep13) [inline src/Cake/Composition/ModuleLoader.cs]: Need a space after closing }

File: src/Cake/Composition/ModuleLoader.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        }
+
+        private DirectoryPath GetModulePath(CakeOptions options)
+        {
+            var root = _configuration.GetValue(Constants.Paths.Modules);
+            if (string.IsNullOrWhiteSpace(root))
+            {
+                return options.Script.GetDirectory().Combine("tools/modules");
+            }
```
Comment (gep13) [inline src/Cake/Composition/ModuleLoader.cs]: Need a space after }

File: src/Cake/Composition/ModuleLoader.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        private void RegisterExternalModules(IEnumerable<Type> moduleTypes, ILifetimeScope scope)
+        {
+            var builder = new ContainerBuilder();
+            foreach (var moduleType in moduleTypes)
+            {
+                _log.Debug("Registering module {0}...", moduleType.FullName);
+                builder.RegisterType(moduleType).As<ICakeModule>().SingleInstance();
+            }
```
Comment (gep13) [inline src/Cake/Composition/ModuleLoader.cs]: Need a space after }

File: src/Cake/Composition/ModuleSearcher.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                AppDomain.CurrentDomain.ReflectionOnlyAssemblyResolve += CurrentDomain_ReflectionOnlyAssemblyResolve;
+                var assembly = Assembly.ReflectionOnlyLoadFrom(path.FullPath);
+                var attributes = CustomAttributeData.GetCustomAttributes(assembly);
+                foreach (var attribute in attributes)
+                {
+                    if (attribute.AttributeType.FullName == typeof(CakeModuleAttribute).FullName)
+                    {
+                        return true;
+                    }
+                }
```
Comment (gep13) [inline src/Cake/Composition/ModuleSearcher.cs]: Need a space after }

File: src/Cake.Core/Modules/CoreModule.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+using Cake.Core.Scripting.Analysis;
+using Cake.Core.Tooling;
+
+namespace Cake.Core.Modules
+{
+    /// <summary>
+    /// The module responsible for registering
+    /// default types in the Cake.Core assembly.
+    /// </summary>
+    public sealed class CoreModule : ICakeModule
```
Comment (gep13) [inline src/Cake.Core/Modules/CoreModule.cs]: Should this be CakeCoreModule?

File: src/Cake/Modules/ArgumentsModule.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿using Cake.Core;
+using Cake.Core.Composition;
+
+namespace Cake.Modules
+{
+    internal sealed class ArgumentsModule : ICakeModule
```
Comment (gep13) [inline src/Cake/Modules/ArgumentsModule.cs]: Should this be CakeArgumentsModule?

File: src/Cake/Modules/ConfigurationModule.cs [hunk 8]
```diff
@@ -1,10 +1,11 @@
 ﻿using System.Diagnostics;
 using Autofac;
+using Cake.Core.Composition;
 using Cake.Core.Configuration;
 
-namespace Cake.Autofac
+namespace Cake.Modules
 {
-    internal sealed class ConfigurationModule : Module
+    internal sealed class ConfigurationModule : ICakeModule
```
Comment (gep13) [inline src/Cake/Modules/ConfigurationModule.cs]: Should this be CakeConfigurationModule?

File: src/Cake/Modules/ScriptingModule.cs [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 using Cake.Scripting.Mono;
 using Cake.Scripting.Roslyn;
 using Cake.Scripting.Roslyn.Nightly;
 using Cake.Scripting.Roslyn.Stable;
 
-namespace Cake.Autofac
+namespace Cake.Modules
 {
-    internal sealed class ScriptingModule : Module
+    internal sealed class ScriptingModule : ICakeModule
```
Comment (gep13) [inline src/Cake/Modules/ScriptingModule.cs]: Should this be CakeScriptingModule?

**Comment by gep13** [comment]
@patriksvensson this seems to work to me :smile:

image

Left a couple quick comments, but nothing major, if you can get them fixed up, and if @devlead agrees, we can get this merged, and the release process started.

**Comment by RichiCoder1** [comment]
I'm curious why you didn't use something like MEF, if only for the initial module loading?

**Comment by patriksvensson** [comment]
@RichiCoder1 Didn't see a good reason to use MEF for this to be honest.

**Comment by patriksvensson** [comment]
@RichiCoder1 Also not a big fan of MEF for IoC and wanted the abstractions to be close to what already exist in Cake's internals.

**Comment by devlead** [comment]
@patriksvensson @gep13 LGTM :+1:

**Comment by devlead** [comment]
@patriksvensson rebase and we should be good to know.

**Comment by patriksvensson** [comment]
@devlead Squashed and rebased.

**Comment by patriksvensson** [comment]
@cake-build/team Whaaat? No thank you? :laughing:

**Comment by gep13** [comment]
@patriksvensson ooops, you are of course correct...

Thank you @devlead for your epic work here! :stuck_out_tongue:

**Pull request merged**